### ISSUE TITLE
Calling subscribe and unsubscribe methods multiple times 

### ISSUE THREAD

**Issue by rokumatsumoto** [mclintprojects/actioncable-vue#4]
**Describe the bug**
I have a download button component on a web page. This page will be the most visited page. You can think it's a product page so I need to use my resources very efficient.

When a user clicks the download button, the server responds with a URL or calls the download service (trigger a background job) and responds with 200. A spinner icon appears on the download button, so the user understands he/she needs to wait.

User can click download button multiple times if he/she wants.

I want to initiate the action cable process on demand so I call the below code on download button click function.

`this.$cable.subscribe({ channel: 'DownloadChannel', id: this.recordId });`

After download service completes its job, it broadcasts to `DownloadChannel` with a URL.

On received function, I call the below code and the user starts downloading the file.

`this.$cable.unsubscribe('DownloadChannel');`
`window.location = data.url;`

After the second click on the download button, an error appears.

**To Reproduce**
Please see [download_button.vue](https://gist.github.com/rokumatsumoto/7fc11afa8300484a358f8ffd2267f0e2) component for testing purpose

1. debug: true,
    debugLevel: 'all'
2. Click download button 
`Successfully connected to channel 'DownloadChannel'.`
`Unsubscribing from channel 'DownloadChannel'.`
3. Click download button again
`Uncaught TypeError: Cannot read property '_uid' of undefined`

**Expected behavior**
User can click download button multiple times if he/she wants.
Calling `unsubscribe` method removes the `DownloadChannel` from `this._channels` object.
https://github.com/mclintprojects/actioncable-vue/blob/b1d57527c65ff3484a143c3deb70442d27a7d874/src/cable.js#L103
`_addChannel` invoke only once when the component mounted.
https://github.com/mclintprojects/actioncable-vue/blob/b1d57527c65ff3484a143c3deb70442d27a7d874/src/cable.js#L176
If the `subscribe` method can invoke the `_addChannel` method then my problem will resolve. 

**Screenshots**
After the second click on the download button, an error appears on this line.
`Uncaught TypeError: Cannot read property '_uid' of undefined`
https://github.com/mclintprojects/actioncable-vue/blob/b1d57527c65ff3484a143c3deb70442d27a7d874/src/cable.js#L201

**Variables on this line after first click**
![Screen Shot 2019-10-17 at 15 14 09](https://user-images.githubusercontent.com/28678254/67007824-e49fb580-f0f0-11e9-9ef2-3d86b2153da2.png)

**Variables on this line after second click**
![Screen Shot 2019-10-17 at 15 08 39](https://user-images.githubusercontent.com/28678254/67007527-47448180-f0f0-11e9-9d8f-16fb4d7fe7e5.png)

**Websocket network after second click**
![Screen Shot 2019-10-17 at 17 09 08](https://user-images.githubusercontent.com/28678254/67016440-dd80a380-f100-11e9-99cd-0cab3b51f3a1.png)

**Plugin version (please complete the following information):**
`"actioncable-vue": "1.3.0",`

**Comment by rokumatsumoto** [comment]
Fix for now (Code smell)

``` download_button.js
    addDownloadChannel() {
      const downloadChannel = Object.entries(this.$options.channels)[0];
      this.$cable._addChannel(downloadChannel[0], downloadChannel[1], this);
    },
    download() {
      this.addDownloadChannel();
      this.$cable.subscribe({ channel: 'DownloadChannel', id: this.recordId });

      // after receive download url from server -> unsubscribe DownloadChannel
      // calling `unsubscribe` here for demonstrating the problem
      const vm = this;
      setTimeout(function(){
        vm.$cable.unsubscribe('DownloadChannel');
       }, 1000);
    }
```

**Comment by mclintprojects** [comment]
Hey, @rokumatsumoto! I have only two questions to help ascertain that the fix actually needs to happen in actioncable-vue. 

Is your download channel broadcasting the download URL back to a specific user on your product page or does it broadcast the download URL to the all the users currently on the product page and the user determines if the download URL is for them? If yes to the latter, is that reason why you need to use resources efficiently?

Here's some important context for the fix to your issue: actioncable-vue invokes the `_addChannel` during the component's mount hook so that it can get access to the execution context of your component. Without that execution context it'd be impossible to use `this.` to access the computed properties, data, methods etc in your component. To get that execution context when you call `this.$cable.subscribe({})` would mean that the majority of users will have to explicitly pass in a `this` parameter and it doesn't make for a nice API.

**Comment by rokumatsumoto** [comment]
Hi, @mclintprojects! Download URL is available for all registered users. URL is the same for all users who want to download product info. In the following scenarios, users will subscribe to `DownloadChannel` when they click the download button.

1. A newly created product has not yet had a download URL. If the user clicks the download button on that product page, download service will be invoked, generated URL will be saved to the database when the zip process completed and the URL will send back to the user via **Action Cable**.

2. If the download link has already been generated, the URL will fetch from the database and send it back to the user via **XHR**. 

3. If the download link has not already been generated, multiple users may click the download button at the same time. 
The user clicking the button first starts the zip creation process and waits for the URL to be passed from **Action Cable**. Other users who click the button later will see that the zip creation process has started and likewise wait for the URL to be passed from **Action Cable**.

> Is your download channel broadcasting the download URL back to a specific user on your product page?

Only when the URL has not been generated. Like scenarios 1 and 3.

> or does it broadcast the download URL to the all the users currently on the product page 

Only when the URL has not been generated. Like scenario 3.

> and the user determines if the download URL is for them? 

URL is the same for all users who want to download product info. If the download link has already been generated, the URL will fetch from the database and send it back to the user via **XHR**.  (On success callback, user will unsubscribe the `DownloadChannel`)

> If yes to the latter, is that reason why you need to use resources efficiently?

I initiated (subscribe) the Action Cable process on-demand (clicking download button) and terminated (unsubscribe) the Action Cable process when the URL sent back to the user. 

Should I disconnect the client from Action Cable when they unsubscribed? (for resource efficiency) If your answer yes, how can I do that on actioncable-vue?

> To get that execution context when you call this.$cable.subscribe({}) would mean that the majority of users will have to explicitly pass in a this parameter and it doesn't make for a nice API.

I understand, can you add an option to support my scenario?

**Comment by mclintprojects** [comment]
Hey, @rokumatsumoto! Thanks for getting back to me. I now understand what you want to do. Your use-case made me realize that I never even added a way to re-subscribe if a user intentionally unsubscribed. I'll add it in a update soon.

**Comment by mclintprojects** [comment]
Hi, @rokumatsumoto. This should be fixed now. Update to v1.4.1 and let me know how it goes.

**Comment by rokumatsumoto** [comment]
Hi, @mclintprojects. Sorry for the late response. I updated the library to v1.4.1, now it's working seamlessly.

By the way, I'm using this config.

```
debug: true,
debugLevel: 'all',
```

When using 1.3.0, I was seeing these log messages in the console. 

```
Successfully connected to channel 'DownloadChannel'.
Unsubscribing from channel 'DownloadChannel'.
```

Now I see these messages.

```
Successfully connected to channel 'DownloadChannel'.
Message received on channel 'DownloadChannel'.
```
Should the log messages also record the `unsubscribe` method?

**Comment by mclintprojects** [comment]
Hi, @rokumatsumoto! I just realized I forgot to add a log statement when unsubscribing. I'll fix it and publish the patch on NPM as soon as I can.

**Comment by bruno-b-martins** [comment]
Thank you for the fix! I was wondering why couldn't I re-subscribe to a channel and starting to dig into your code 😝

**Comment by rokumatsumoto** [comment]
Thanks, @mclintprojects.

Should I disconnect the client from Action Cable when they unsubscribed? (for resource efficiency) If your answer yes, how can I do that on actioncable-vue?

**Comment by mclintprojects** [comment]
@rokumatsumoto I just published a patch (v1.4.2) that adds the log statement when you unsubscribe. 

No, you don't need to. If you unsubscribe from a channel, the resources you'll still be using are negligible. Think of it as if you have an iron plugged into a socket that's turned off. You're not wasting any electricity. Also, it's probably quicker to get back to using the iron because it's already plugged into the socket.
